### PR TITLE
connect/disconnect reducers called when executing reducer from from CLI

### PR DESCRIPTION
**Pull request by jdetter**
# Description of Changes

 - When you invoke a reducer from the CLI, now OnConnect and OnDisconnect are invoked in the module.

# API and ABI

 - [ ] This is a breaking change to the module ABI
 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API
 - [x] This is a change in behavior that may be API breaking

*If the API is breaking, please state below what will break*

### CODE REVIEW

File: crates/client-api/src/routes/database.rs [hunk 1]
```diff
@@ -92,23 +92,37 @@ pub async fn call<S: ControlStateDelegate + NodeDelegate>(
...
-                    log::debug!("Attempt to call reducer with invalid arguments");
-                    StatusCode::BAD_REQUEST
-                }
-                ReducerCallError::NoSuchModule(_) => StatusCode::NOT_FOUND,
-                ReducerCallError::NoSuchReducer => {
-                    log::debug!("Attempt to call non-existent reducer {}", reducer);
-                    StatusCode::NOT_FOUND
+
+    let result = if let Err(e) = module.call_identity_connected_disconnected(caller_identity, true).await {
+        return Err((StatusCode::NOT_FOUND, format!("{:#}", anyhow::anyhow!(e))).into());
```
Comment (gefjon) [inline crates/client-api/src/routes/database.rs:97]: I'm not sure `StatusCode::NOT_FOUND` is appropriate here, tho I don't really know what will cause `call_identity_connected_disconnected` to return `Err`.

File: crates/client-api/src/routes/database.rs [hunk 2]
```diff
@@ -92,23 +92,37 @@ pub async fn call<S: ControlStateDelegate + NodeDelegate>(
...
+    } else {
```
Comment (gefjon) [inline crates/client-api/src/routes/database.rs:98]: This `if`/`else` chain is kinda grody, and I don't think it's necessary. Can this be:

```rust
if let Err(e) = module.call_identity_connected_disconnected(caller_identity, true).await {
    // handle error
}
let result = match module.call_reducer(whatever).await {
    // whatever
};
```

It also seems like the `if let Err(e) = foo { return Err(whatever); }` could be made into a helper function, something like:

```rust
fn with_status_code(code: StatusCode, err: impl std::string::ToString) -> impl Axum::IntoResponse {
    (code, err.to_string())
}

// handling an error
let result = module.call_reducer(caller_identity, None, &reducer, args)
    .map_err(|e| with_status_code(StatusCode::NOT_FOUND, e)?;
```

File: crates/client-api/src/routes/database.rs [hunk 3]
```diff
@@ -92,23 +92,37 @@ pub async fn call<S: ControlStateDelegate + NodeDelegate>(
...
+                    .call_identity_connected_disconnected(caller_identity, false)
+                    .await
+                {
+                    return Err((StatusCode::NOT_FOUND, format!("{:#}", anyhow::anyhow!(e))).into());
+                } else {
+                    rcr
                 }
-            };
+            }
+            Err(e) => {
```
Comment (gefjon) [inline crates/client-api/src/routes/database.rs:110]: Should we call disconnected somewhere in this block?

File: crates/client-api/src/routes/database.rs (hunk 1 above)
Comment (jdetter) [reply crates/client-api/src/routes/database.rs:97]: It appears to only return an error on `ModuleNotFound`, I figured this was appropriate 🤷‍♂️

**Review by gefjon — APPROVED** [review]

**Pull request merged**